src/ec2_gha/start.py

```diff
@@ -250,45 +388,114 @@ def create_instances(self) -> dict[str, str]:
+            action_sha = resolve_ref_to_sha(action_ref)
 
             user_data_params = {
+                "action_sha": action_sha,  # The resolved SHA
                 "cloudwatch_logs_group": self.cloudwatch_logs_group,
+                "debug": self.debug,
                 "github_workflow": environ.get("GITHUB_WORKFLOW", ""),
                 "github_run_id": environ.get("GITHUB_RUN_ID", ""),
                 "github_run_number": environ.get("GITHUB_RUN_NUMBER", ""),
                 "homedir": self.home_dir,
-                "labels": labels,
+                "instance_name": instance_name_value,  # Add the generated instance name
                 "max_instance_lifetime": self.max_instance_lifetime,
                 "repo": self.repo,
                 "runner_grace_period": self.runner_grace_period,
                 "runner_initial_grace_period": self.runner_initial_grace_period,
+                "runner_poll_interval": self.runner_poll_interval,
+                "runner_registration_timeout": environ.get("INPUT_RUNNER_REGISTRATION_TIMEOUT", "").strip() or RUNNER_REGISTRATION_TIMEOUT,
                 "runner_release": self.runner_release,
+                "runners_per_instance": str(self.runners_per_instance),
+                "runner_tokens": runner_tokens,  # Space-delimited tokens
+                "runner_labels": runner_labels,  # Pipe-delimited labels
                 "script": self.script,
                 "ssh_pubkey": self.ssh_pubkey,
-                "token": token,
                 "userdata": self.userdata,
             }
-            params = self._build_aws_params(user_data_params)
-            if self.root_device_size > 0:
+            params = self._build_aws_params(user_data_params, idx=idx)
+            if self.root_device_size != "0":
                 params = self._modify_root_disk_size(ec2, params)
-            result = ec2.run_instances(**params)
+
+            # Check UserData size before calling AWS
+            user_data_size = len(params.get("UserData", ""))
+            if user_data_size > 16384:
+                raise ValueError(
+                    f"UserData exceeds AWS limit: {user_data_size} bytes (limit: 16384 bytes, "
+                    f"over by: {user_data_size - 16384} bytes). "
+                    f"Template needs to be reduced by at least {user_data_size - 16384} bytes."
+                )
+
+            try:
+                result = ec2.run_instances(**params)
+            except Exception as e:
+                if "User data is limited to 16384 bytes" in str(e):
+                    # This shouldn't happen if our check above works, but just in case
+                    raise ValueError(
+                        f"UserData exceeds AWS limit: {user_data_size} bytes (limit: 16384 bytes, "
+                        f"over by: {user_data_size - 16384} bytes)"
+                    ) from e
+                raise
```

Comment:
🐑 this seems like a lot; maybe remove this pre- and post- check and let the exception get raised by run_instances? Or perhaps catch + re-raise with `add_note` about reducing template size if userdata is too large?